pkg/conn/basedb.go

```diff
@@ -0,0 +1,131 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package conn
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"sync"
+
+	"github.com/pingcap/dm/dm/config"
+	"github.com/pingcap/dm/pkg/retry"
+	"github.com/pingcap/dm/pkg/terror"
+)
+
+// DBProvider providers BaseDB instance
+type DBProvider interface {
+	Apply(config config.DBConfig) (*BaseDB, error)
+}
+
+type defaultDBProvider struct {
+}
+
+// DefaultDBProvider is global instance of DBProvider
+var DefaultDBProvider DBProvider
+
+func init() {
+	DefaultDBProvider = &defaultDBProvider{}
+}
+
+// Apply will build BaseDB with DBConfig
+func (d *defaultDBProvider) Apply(config config.DBConfig) (*BaseDB, error) {
+	dsn := fmt.Sprintf("%s:%s@tcp(%s:%d)/?charset=utf8mb4&interpolateParams=true&maxAllowedPacket=%d",
+		config.User, config.Password, config.Host, config.Port, *config.MaxAllowedPacket)
+
+	var maxIdleConns int
+	rawCfg := config.RawDBCfg
+	if rawCfg != nil {
+		if rawCfg.ReadTimeout != "" {
+			dsn += fmt.Sprintf("&readTimeout=%s", rawCfg.ReadTimeout)
+		}
+		if rawCfg.WriteTimeout != "" {
+			dsn += fmt.Sprintf("&writeTimeout=%s", rawCfg.WriteTimeout)
+		}
+		maxIdleConns = rawCfg.MaxIdleConns
+	}
+	db, err := sql.Open("mysql", dsn)
+	if err != nil {
+		return nil, terror.ErrDBDriverError.Delegate(err)
```

Comment:
```suggestion
		return nil, terror.DBErrorAdapt(err, terror.ErrDBDriverError)
```
use adaptor is better, we will get more concrete error, such as `ErrDBBadConn`